### PR TITLE
Build version 3.22.11 with qt515

### PR DESCRIPTION
**Pull request by izahn**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by izahn** [comment]
@conda-forge-admin please rerender

**Comment by izahn** [comment]
@gillins we're getting

```
[1323/5199] Linking CXX shared library output\bin\qgis_core.dll
FAILED: output/bin/qgis_core.dll src/core/qgis_core.lib 
cmd.exe /C "cd . && %BUILD_PREFIX%\Library\bin\cmake.exe -E vs_link_dll --intdir=src\core\CMakeFiles\qgis_core.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\qgis_core.rsp  /out:output\bin\qgis_core.dll /implib:src\core\qgis_core.lib /pdb:output\bin\qgis_core.pdb /dll /version:3.26 /machine:x64 /INCREMENTAL:NO  && cd ."
LINK: command "C:\PROGRA~2\MICROS~1\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\qgis_core.rsp /out:output\bin\qgis_core.dll /implib:src\core\qgis_core.lib /pdb:output\bin\qgis_core.pdb /dll /version:3.26 /machine:x64 /INCREMENTAL:NO /MANIFEST /MANIFESTFILE:output\bin\qgis_core.dll.manifest" failed (exit code 1120) with the following output:
   Creating library src\core\qgis_core.lib and object src\core\qgis_core.exp
vector_tile.pb.cc.obj : error LNK2019: unresolved external symbol "class google::protobuf::internal::ExplicitlyConstructed<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,8> google::protobuf::internal::fixed_address_empty_string" (?fixed_address_empty_string@internal@protobuf@google@@3V?$ExplicitlyConstructed@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@$07@123@A) referenced in function "private: static class vector_tile::Tile_Value * __cdecl google::protobuf::Arena::InternalHelper<class vector_tile::Tile_Value>::Construct<class google::protobuf::Arena *>(void *,class google::protobuf::Arena * &&)" (??$Construct@PEAVArena@protobuf@google@@@?$InternalHelper@VTile_Value@vector_tile@@@Arena@protobuf@google@@CAPEAVTile_Value@vector_tile@@PEAX$$QEAPEAV123@@Z)
vector_tile.pb.cc.obj : error LNK2019: unresolved external symbol "struct std::atomic<bool> google::protobuf::internal::init_protobuf_defaults_state" (?init_protobuf_defaults_state@internal@protobuf@google@@3U?$atomic@_N@std@@A) referenced in function "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl google::protobuf::internal::GetEmptyString(void)" (?GetEmptyString@internal@protobuf@google@@YAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
output\bin\qgis_core.dll : fatal error LNK1120: 2 unresolved externals
```
on the windows builds. It looks like you fixed this (or something very similar) once before in https://github.com/qgis/QGIS/pull/41130 , any chance you can help fix it again here?

**Comment by gillins** [comment]
Sorry for slow response @izahn. I think you are on the right track. We originally added this fix via a patch here: https://github.com/conda-forge/qgis-feedstock/pull/142 and then up-streamed here: https://github.com/qgis/QGIS/pull/41130 but sadly it seems to have been removed again. I think the QGIS people are mainly interested on static linking on Windows, would be good if they added a CI for conda-forge builds like GDAL does...

Anyway, I wonder if you add a patch like the original (https://github.com/conda-forge/qgis-feedstock/pull/142/commits/e2bb9e2206967b16b3a48c7d4b2abc435a54e2a0#diff-02aa49ce3834d09fd28cd0d23bc022a9faacebcd19e86cbba7d61285d573d2ea) that might work?

**Comment by izahn** [comment]
I tried patching in https://github.com/conda-forge/qgis-feedstock/pull/253/commits/059612df84a710c1ddc2acd2b9b67cdac1b1e3f0 so far without success. So much has changed that I'm not really sure what needs to be done and I don't really know cmake very well.

File: recipe/0001-use-protobuf-dlls.patch [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+diff --git a/src/core/CMakeLists.txt b/src/core/CMakeLists.txt
+index 7537413..8eed55d 100644
+--- a/src/core/CMakeLists.txt
++++ b/src/core/CMakeLists.txt
+@@ -1813,6 +1813,10 @@ if (NOT MSVC)
+   set_source_files_properties(${VECTOR_TILE_PROTO_SRCS} PROPERTIES COMPILE_FLAGS -w)
+ endif()
+ 
++if (MSVC)
++  set_source_files_properties(${VECTOR_TILE_PROTO_SRCS} PROTOBUF_USE_DLLS)
```
Comment (gillins) [inline recipe/0001-use-protobuf-dlls.patch:19]: Should be:
```
set_source_files_properties(${VECTOR_TILE_PROTO_SRCS} vectortile/qgsvectortilewriter.cpp PROPERTIES COMPILE_DEFINITIONS PROTOBUF_USE_DLLS)
```

Comment (izahn) [reply recipe/0001-use-protobuf-dlls.patch]: I see you fixed this in https://github.com/conda-forge/qgis-feedstock/pull/253/commits/82a74382cbdd911eec68f7a359ab4c7f5f4ae006 thank you!

Comment (gillins) [reply recipe/0001-use-protobuf-dlls.patch]: Sorry - I think I gave you the wrong information! I'll try and get this upstreamed (again).

**Comment by izahn** [comment]
We are almost to the finish line here. OSX build pass, Linux builds fail when run in docker but succeed when built locally and can be manually uploaded, and Windows py38 build passes. 

That leaves Windows py39 and py310, which area failing with

```

    import qgis.PyQt.uic.pyuic

  File "%SRC_DIR%\build\output\python\qgis\__init__.py", line 76, in <module>

    os.add_dll_directory(p)

  File "%PREFIX%\lib\os.py", line 1117, in add_dll_directory

    cookie = nt._add_dll_directory(path)

OSError: [WinError 87] The parameter is incorrect: '.'
```

I don't have access to a windows machine which makes debugging harder for me, maybe someone with a windows machine can try building it locally and investing the error?

**Comment by izahn** [comment]
Manual linux builds are available:

- linux_64_python3.8
  - package: https://anaconda.org/izahn/qgis/3.22.11/download/linux-64/qgis-3.22.11-py38hb7d1217_0.tar.bz2
  - log: https://www.dropbox.com/s/xd9np31v8fdw4fb/linux_64_python3.8.____cpython.log?dl=0
- linux_64_python3.9
  - package: https://anaconda.org/izahn/qgis/3.22.11/download/linux-64/qgis-3.22.11-py39h3a22410_0.tar.bz2
  - log: https://www.dropbox.com/s/402eajcarap8dtx/linux_64_python3.9.____cpython.log?dl=0
- linux_64_python3.10
  - package: https://anaconda.org/izahn/qgis/3.22.11/download/linux-64/qgis-3.22.11-py310heaafcf0_0.tar.bz2
  - log: https://www.dropbox.com/s/asksmjar2k1skrb/linux_64_python3.10.____cpython.log?dl=0

**Comment by gillins** [comment]
Now getting:
```
(%PREFIX%) %SRC_DIR%>%PREFIX%\python.exe -c "import qgis.core" 
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\bld\qgis_1663715127268\_test_env\Library\python\qgis\core\__init__.py", line 25, in <module>
    from qgis._core import *
ImportError: DLL load failed while importing _core: The specified procedure could not be found.
```

Is there anyway to grab the build so I can run some tests?

**Comment by izahn** [comment]
I turned on artifact storage, builds from the last run are at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=571767&view=artifacts&pathAsName=false&type=publishedArtifacts

Strange that the win py38 version doesn't have this problem but the other two do.

**Comment by gillins** [comment]
Thanks @izahn

**Comment by SrNetoChan** [comment]
@izahn and @gillins thank you for all your effort! I really miss using qgis from conda

**Comment by gillins** [comment]
@conda-forge-admin, please rerender

**Comment by gillins** [comment]
OK I think we are all good now. I'm still very puzzled by the Linux/docker issue. I'm sure this must be related to memory use as those `sip` generated files are enormous, but don't really have anything concrete to offer....

**Comment by SrNetoChan** [comment]
@izahn and @gillins thank you for all your effort! I really miss using qgis from conda.

What does this mean for Linux? It won't be available? It needs manual upload? Or maybe the builds will pass after the merge?

**Comment by izahn** [comment]
> @izahn and @gillins thank you for all your effort! I really miss using qgis from conda.
> 
> What does this mean for Linux? It won't be available? It needs manual upload? Or maybe the builds will pass after the merge?

Linux builds need to be manually uploaded. They are ready to go (see https://github.com/conda-forge/qgis-feedstock/pull/253#issuecomment-1252390962 ) just need someone to upload them to the conda-forge channel.

**Comment by gillins** [comment]
Unfortunately, I don't have permissions to upload to the conda-forge channel... @ocefpaf can you do this if we merge?

**Comment by izahn** [comment]
> Unfortunately, I don't have permissions to upload to the conda-forge channel... @ocefpaf can you do this if we merge?
> 
> Also: are we going to have to do this each time there is a migration?

Yes, until whatever causes the linux builds to fail is resolved.

**Comment by gillins** [comment]
That's going to be a pain. Hopefully resolved soon. @izahn do you want to add yourself as a maintainer?

**Comment by izahn** [comment]
> That's going to be a pain. Hopefully resolved soon.

Yes, but not that unusual and there is a process for it: https://github.com/conda-forge/cfep/blob/main/cfep-03.md

> @izahn do you want to add yourself as a maintainer?

Yes I will do that, thanks.

**Comment by izahn** [comment]
@conda-forge-admin please rerender

**Comment by SrNetoChan** [comment]
@ocefpaf do you have any ideas for the issue @izahn is facing with the docker builds on linux? They seem to work fine locally, but fail in the CI.

**Comment by SrNetoChan** [comment]
Anyway +1 for merging as it is, even if the Linux builds are failing and we can't upload them manually. At least windows and mac users will be able to use qgis from conda again.

And we can try to keep the pace with qgis stable versions

**Comment by ocefpaf** [comment]
> @ocefpaf do you have any ideas for the issue @izahn is facing with the docker builds on linux? They seem to work fine locally, but fail in the CI.

No idea. Let me investigate this a bit before we resort to manual uploads. Looking into it now...

**Comment by ocefpaf** [comment]
The build did not work for me locally, see https://gist.github.com/ocefpaf/6d4998c710bc2282e1a8f27bf06364f2.

@izahn are you calling `python build-locally.py ` on the feedstock? If not you may be using some local libraries and that may be the reason why this is passing for you.

**Comment by izahn** [comment]
> @izahn are you calling `python build-locally.py ` on the feedstock? 

No, building in docker fails locally just as it does on the CI. I have to run `mamba build -m ...`, no docker involved.

> If not you may be using some local libraries and that may be the reason why this is passing for you.

Possible, but there is no error message to suggest that. I suspect (without evidence) that there is some resource limit in docker that causes the problem. That's just my wild speculation though.

**Comment by ocefpaf** [comment]
> No, building in docker fails locally just as it does on the CI. I have to run `mamba build -m ...`, no docker involved.

The CI does use our docker image. You can reproduce the same build with `python build-locally.py 2>&1 | tee log`. I'm clueless here but I suspect it has to do with our complex dependencies needed to build qgis.

**Comment by SrNetoChan** [comment]
> The build did not work for me locally, see https://gist.github.com/ocefpaf/6d4998c710bc2282e1a8f27bf06364f2.
> 
> @izahn are you calling `python build-locally.py ` on the feedstock? If not you may be using some local libraries and that may be the reason why this is passing for you.

@nyalldawson @jef-n sorry to bother you. Does any of you have tips on what could be happening here? 

Everything seems finally to be compiling fine with Qt 5.15 in mac and windows but it's failing on linux for all versions of Python.

Thanks

**Comment by gillins** [comment]
Hi @ocefpaf any more thoughts on this? Keen to push on since it has taken such a long time to get to this point, even if we have to do the manual upload....

**Comment by izahn** [comment]
Yeah, the linux situation is weird and I don't fully understand it either. Here is what I know:

1. conda-forge CI builds **fail**
2. `build-locally.py` **fails**
3. `mamba build -m .ci_support/...` on virtualbox starting with Centos7 minimal **fails**
4. `mamba build -m .ci_support/...` on our compute cluster running RHEL7 **succeeds**
5. `mamba build -m .ci_support/...`  on my laptop running Arch Linux **succeeds**

Possible next steps are a) installing Centos minimal on physical hardware and try there, and b) getting the package list installed on the cluster and installing all that stuff in virtualbox and test that. Both are labor intensive and I haven't had time to try. Any and all suggestions and help appreciated.

**Comment by ocefpaf** [comment]
> Both are labor intensive and I haven't had time to try. Any and all suggestions and help appreciated.

Indeed. And I also think it is not necessary. It is probably something that is right in front of us but we can missing. I need to familiarize myself again with the dependency stack to see what changed. My guess is that something in the dependency tree changed and is causing the errors.

**Comment by SrNetoChan** [comment]
In @izahn machine that is working, is it possible to check from where all dependencies are coming when mamba is building. To see if any dependency comes from outside conda environment?

**Comment by izahn** [comment]
The full log from a successful RHEL7 build is at https://www.dropbox.com/s/wtncifj33465fhg/linux_64_python3.10._RHEL7_cpython.log?dl=0

**Comment by gillins** [comment]
> At this point I'm tempted to upload @izahn's builds 

I agree, would be good to get this moving since `qgis` has been stuck on an old version for a long time. Skipping Linux for now could be another option, although even less ideal.

**Comment by ocefpaf** [comment]
OK. Merging this to "cash-in" the passing ones and I'll discuss today with the core group the uploading of the Linux builds. Thanks all for your patience and @izahn for the Linux builds.

**Comment by ocefpaf** [comment]
Annnnnd, everything is failing on main b/c of the new SIP :-/

**Comment by izahn** [comment]
> Annnnnd, everything is failing on main b/c of the new SIP :-/

Bad timing!

**Comment by ocefpaf** [comment]
The linux builds are up though :smile: 

What do you folks want to do here? Fix it or pin to an older sip?

**Comment by izahn** [comment]
> The linux builds are up though

🎊 

> 
> What do you folks want to do here? Fix it or pin to an older sip?

Looking at the log it seems that `sip 6.6.2` was used in both cases. If we could just pin something temporarily and get the builds up I would be for that, but I don't think we know what to pin in order to get things to build as they did in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=575723&view=results . Ugh.

**Pull request merged**